File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.kt
/*
두 큐 합 같게 만들기 / 코틀린

두 큐의 원소 합이 같도록 작업에 필요한 최소 횟수

 */
class Solution {
    fun solution(queue1: IntArray, queue2: IntArray): Int {
        var answer: Int = 0
        var size : Int = queue1.size + queue2.size
        var oneSum : Long = queue1.sumOf { it.toLong() }
        var twoSum : Long = queue2.sumOf { it.toLong() }
        var q1 : ArrayDeque<Int> = ArrayDeque(queue1.map { it.toInt() }.toList())
        var q2 : ArrayDeque<Int> = ArrayDeque(queue2.map { it.toInt() }.toList())

        while(answer <= size * 2){
            var tmp : Int = 0
            if(oneSum == twoSum) return answer
            if(oneSum > twoSum) {
                tmp = q1.first()
                q1.removeFirst()
                oneSum -= tmp
                twoSum += tmp
                q2.addLast(tmp)
                answer++
            }else{
                tmp = q2.first()
                q2.removeFirst()
                oneSum += tmp
                twoSum -= tmp
                q1.addLast(tmp)
                answer++
            }
        }
        return -1
    }
}

File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.cpp
/*
두 큐 합 같게 만들기

그리디로 음 작은것에 대해서 더해주는 방식을 반복하면될듯?

*/

#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> queue1, vector<int> queue2) {
    int answer = 0;
    
    long long oneSum = 0;
    long long twoSum = 0;
    queue<int> q1;
    queue<int> q2;
    
    for(int a : queue1){
        q1.push(a);
        oneSum += a;
    }
    
    for(int a : queue2) {
        q2.push(a);
        twoSum += a;
    }
    
    while(answer <= (q1.size() + q2.size()) * 2){
        if(oneSum == twoSum) return answer;
        
        answer++;
        
        if(oneSum < twoSum){
            int tmp = q2.front();
            oneSum += tmp;
            twoSum -= tmp;
            q2.pop();
            q1.push(tmp);
        }else {
            int tmp = q1.front();
            oneSum -= tmp;
            twoSum += tmp;
            q1.pop();
            q2.push(tmp);
        }
    }
    
    return -1;
}